src/layers/icmp.rs

```diff
@@ -0,0 +1,107 @@
+//! ICMP Datagram
+
+use std::convert::TryInto;
+
+use serde::Serialize;
+
+use crate::errors::Error;
+use crate::layers::ipv4;
+use crate::types::IPv4Address;
+use crate::Layer;
+
+/// IANA Assigned protocol number for ICMP
+pub const IPPROTO_ICMP: u8 = 1_u8;
+
+// Register ICMP with Protocol Handler in IPv4
+pub(crate) fn register_defaults() -> Result<(), Error> {
+    ipv4::register_protocol(IPPROTO_ICMP, ICMP::creator)
+}
+
+#[derive(Default, Debug, Serialize)]
+#[serde(untagged)]
+pub enum IcmpType {
+    #[default]
+    Unused,
+    Echo(IcmpEcho),
```

Comment:
Since `ECHO Request` and `ECHO Reply` are separate packets, it's a good idea to have them as separate variants in an `Enum`.